File: LoginovaBackend/LoginovaAPI/Models/EmpresaCliente.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace LoginovaAPI.Models;

/// <summary>
/// Representa una empresa que compró Loginova (no un cliente de recogidas:
/// esto es el CRM interno del vendedor para llevar control de a quién se le
/// vendió, con qué instalación cuenta y cuándo vence su membresía). Cada
/// empresa cliente corre su propia instalación independiente de Loginova
/// (modelo "una instalación por cliente"); este registro vive únicamente en
/// la instalación del vendedor, nunca en la de sus clientes.
/// </summary>
[Table("empresas_clientes")]
public class EmpresaCliente
{
    [Column("id")]
    public int Id { get; set; }

    [Column("nombre_empresa")]
    public string NombreEmpresa { get; set; } = "";

    [Column("nombre_contacto")]
    public string? NombreContacto { get; set; }

    /// <summary>Teléfono del contacto, en formato internacional (ej. 573001234567) para el link de WhatsApp.</summary>
    [Column("telefono_contacto")]
    public string? TelefonoContacto { get; set; }

    [Column("correo_contacto")]
    public string? CorreoContacto { get; set; }

    /// <summary>URL base de la instalación de Loginova de esta empresa (su propio backend).</summary>
    [Column("url_instalacion")]
    public string? UrlInstalacion { get; set; }

    [Column("fecha_inicio_membresia")]
    public DateTime FechaInicioMembresia { get; set; }

    [Column("fecha_fin_membresia")]
    public DateTime FechaFinMembresia { get; set; }

    /// <summary>Monto que paga la empresa por su membresía (opcional, informativo).</summary>
    [Column("monto_membresia")]
    public decimal? MontoMembresia { get; set; }

    /// <summary>Ej. "Mensual", "Anual", "Único". Libre, no hay catálogo cerrado.</summary>
    [Column("ciclo_pago")]
    public string? CicloPago { get; set; }

    /// <summary>Notas de seguimiento del vendedor/soporte (no visibles para la empresa cliente).</summary>
    [Column("notas")]
    public string? Notas { get; set; }

    /// <summary>Falso cuando se da de baja definitiva a la empresa (no se elimina, para no perder el historial).</summary>
    [Column("activa")]
    public bool Activa { get; set; } = true;

    /// <summary>Última vez que se marcó como enviado un recordatorio manual (WhatsApp, llamada, etc.).</summary>
    [Column("ultimo_recordatorio_enviado")]
    public DateTime? UltimoRecordatorioEnviado { get; set; }

    [Column("fecha_creacion")]
    public DateTime FechaCreacion { get; set; } = DateTime.UtcNow;
}

File: Loginova/lib/models/empresa_cliente.dart
/// Empresa que compró Loginova (CRM interno del vendedor, no un cliente de
/// recogidas). Cada una corre su propia instalación independiente; este
/// modelo solo existe en la instalación del vendedor para llevar control de
/// membresías, contacto y seguimiento.
class EmpresaCliente {
  final int id;
  final String nombreEmpresa;
  final String? nombreContacto;
  final String? telefonoContacto;
  final String? correoContacto;
  final String? urlInstalacion;
  final DateTime fechaInicioMembresia;
  final DateTime fechaFinMembresia;
  final double? montoMembresia;
  final String? cicloPago;
  final String? notas;
  final bool activa;
  final DateTime? ultimoRecordatorioEnviado;
  final DateTime fechaCreacion;
  // Calculados por el backend: "Vencida", "PorVencer" o "Vigente", y los
  // días restantes (negativo si ya venció). Se guardan tal cual en vez de
  // recalcularlos en el cliente para no duplicar la regla de negocio.
  final String estadoMembresia;
  final int diasParaVencimiento;

  EmpresaCliente({
    required this.id,
    required this.nombreEmpresa,
    this.nombreContacto,
    this.telefonoContacto,
    this.correoContacto,
    this.urlInstalacion,
    required this.fechaInicioMembresia,
    required this.fechaFinMembresia,
    this.montoMembresia,
    this.cicloPago,
    this.notas,
    this.activa = true,
    this.ultimoRecordatorioEnviado,
    required this.fechaCreacion,
    required this.estadoMembresia,
    required this.diasParaVencimiento,
  });

  factory EmpresaCliente.fromJson(Map<String, dynamic> json) {
    return EmpresaCliente(
      id: json['id'],
      nombreEmpresa: json['nombreEmpresa'],
      nombreContacto: json['nombreContacto'],
      telefonoContacto: json['telefonoContacto'],
      correoContacto: json['correoContacto'],
      urlInstalacion: json['urlInstalacion'],
      fechaInicioMembresia: DateTime.parse(
        json['fechaInicioMembresia'],
      ).toLocal(),
      fechaFinMembresia: DateTime.parse(json['fechaFinMembresia']).toLocal(),
      montoMembresia: json['montoMembresia']?.toDouble(),
      cicloPago: json['cicloPago'],
      notas: json['notas'],
      activa: json['activa'] ?? true,
      ultimoRecordatorioEnviado: json['ultimoRecordatorioEnviado'] != null
          ? DateTime.parse(json['ultimoRecordatorioEnviado']).toLocal()
          : null,
      fechaCreacion: DateTime.parse(json['fechaCreacion']).toLocal(),
      estadoMembresia: json['estadoMembresia'] ?? 'Vigente',
      diasParaVencimiento: json['diasParaVencimiento'] ?? 0,
    );
  }

  Map<String, dynamic> toRequestJson() {
    return {
      'nombreEmpresa': nombreEmpresa,
      'nombreContacto': nombreContacto,
      'telefonoContacto': telefonoContacto,
      'correoContacto': correoContacto,
      'urlInstalacion': urlInstalacion,
      'fechaInicioMembresia': fechaInicioMembresia.toUtc().toIso8601String(),
      'fechaFinMembresia': fechaFinMembresia.toUtc().toIso8601String(),
      'montoMembresia': montoMembresia,
      'cicloPago': cicloPago,
      'notas': notas,
      'activa': activa,
    };
  }
}
